### PR TITLE
Add Span-based {Try}Parse methods to primitive types

### PR DESCRIPTION
**Pull request by stephentoub**
Adds Parse and TryParse methods to Boolean, Byte, Double, Int16, Int32, Int64, SByte, Single, UInt16, UInt32, UInt64, and Decimal.

Contributes to https://github.com/dotnet/corefx/issues/22403
cc: @jkotas, @KrzysztofCwalina, @tarekgh

### CODE REVIEW

File: src/mscorlib/shared/System/Boolean.cs [hunk 1]
```diff
@@ -178,23 +178,33 @@ public static Boolean Parse(String value)
...
-            // For perf reasons, let's first see if they're equal, then do the
-            // trim to get rid of white space, and check again.
-            if (TrueLiteral.Equals(value, StringComparison.OrdinalIgnoreCase))
+
+            return TryParse(value.AsSpan(), out result);
+        }
+
+        public static bool TryParse(ReadOnlySpan<char> value, out bool result)
+        {
+            ReadOnlySpan<char> trueSpan = TrueLiteral.AsSpan();
```
Comment (tarekgh) [inline src/mscorlib/shared/System/Boolean.cs]: >TrueLiteral.AsSpan() [](start = 42, length = 20)

is this fast operation or should we cache this value in some static?

Comment (jkotas) [inline src/mscorlib/shared/System/Boolean.cs]: It is fast operation, and you cannot cache Span in a static even if you wanted to - Span cannot be assigned to static field.

File: src/mscorlib/shared/System/Byte.cs [hunk 2]
```diff
@@ -111,8 +111,20 @@ public static byte Parse(String s, NumberStyles style, IFormatProvider provider)
...
 
+        public static byte Parse(ReadOnlySpan<char> s, NumberStyles style, IFormatProvider provider)
+        {
+            NumberFormatInfo.ValidateParseStyleInteger(style);
+            return Parse(s, style, NumberFormatInfo.GetInstance(provider));
+        }
+
         private static byte Parse(String s, NumberStyles style, NumberFormatInfo info)
         {
+            if (s == null) ThrowHelper.ThrowArgumentNullException(ExceptionArgument.s);
```
Comment (jkotas) [inline src/mscorlib/shared/System/Byte.cs]: Is there a reason why there is extra private wrapper method in some cases (like here) vs. the code is inlined in other cases, the equivalent change for short is:

```
public static short Parse(String s, NumberStyles style, IFormatProvider provider) 
{ 
    NumberFormatInfo.ValidateParseStyleInteger(style); 
    if (s == null) ThrowHelper.ThrowArgumentNullException(ExceptionArgument.s);  
    return Parse(s.AsSpan(), style, NumberFormatInfo.GetInstance(provider));
}
```

Comment (stephentoub) [inline src/mscorlib/shared/System/Byte.cs]: No reason, other than when editing 12 types with a bazillion overloads, one can lose track of the pattern (especially when they're already inconsistent) :)  I'll review and tweak where necessary.

File: src/mscorlib/shared/System/Boolean.cs (hunk 1 above)
Comment (tarekgh) [inline src/mscorlib/shared/System/Boolean.cs]: thanks for the clarification.

**Comment by tarekgh** [comment]
LGTM.

File: src/mscorlib/shared/System/Byte.cs (hunk 2 above)
Comment (stephentoub) [inline src/mscorlib/shared/System/Byte.cs]: @jkotas, I pushed a commit to make these methods consistent across the primitive types.

**Pull request merged**